ElectricCar.cpp

```diff
@@ -1,16 +1,48 @@
 #include "ElectricCar.hpp"
+
 #include <iostream>
 
+#include "WrongEngine.hpp"
+#include "CarHaveMove.hpp"
+
 ElectricCar::ElectricCar(ElectricEngine* engine)
-    : engine_(engine)
-{
+    : electricEngine_(engine) {
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+ElectricCar::~ElectricCar() {
+    std::cout << __FUNCTION__ << std::endl;
+    delete electricEngine_;
+}
+
+void ElectricCar::restore() {
     std::cout << __FUNCTION__ << std::endl;
+    charge();
+}
+
+void ElectricCar::charge() {
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+int ElectricCar::getBatteryCapacity() const {
+    return electricEngine_->getBatteryCapacity();
+}
+
+int ElectricCar::getElectricPower() const {
+    return electricEngine_->getPower();
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+void ElectricCar::changeEngine(Engine* engine) {
```

Comment:
Can't you just take ElectricEngine as parameter?

Reply:
In this case it wouldn't be a base class function overload.